src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/PosixSignalRegistration.Windows.cs

```diff
@@ -1,18 +1,134 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System;
+using System.Collections.Generic;
+using System.IO;
 
 namespace System.Runtime.InteropServices
 {
-    public sealed class PosixSignalRegistration : IDisposable
+    public sealed partial class PosixSignalRegistration
     {
-        private PosixSignalRegistration() { }
+        private static readonly Interop.Kernel32.ConsoleCtrlHandlerRoutine s_handlerRoutine = HandlerRoutine;
+        private static readonly HashSet<Token> s_handlers = new();
 
-        public static PosixSignalRegistration Create(PosixSignal signal, Action<PosixSignalContext> handler)
-            => throw new PlatformNotSupportedException();
+        private Token? _token;
 
-        public void Dispose()
-            => throw new PlatformNotSupportedException();
+        private PosixSignalRegistration(Token token) => _token = token;
+
+        private static object SyncObj => s_handlers;
+
+        public static partial PosixSignalRegistration Create(PosixSignal signal, Action<PosixSignalContext> handler)
+        {
+            if (handler is null)
+            {
+                throw new ArgumentNullException(nameof(handler));
+            }
+
+            lock (SyncObj)
+            {
+                switch (signal)
+                {
+                    case PosixSignal.SIGINT:
+                    case PosixSignal.SIGQUIT:
+                    case PosixSignal.SIGTERM:
+                    case PosixSignal.SIGHUP:
+                        break;
+
+                    default:
+                        throw new PlatformNotSupportedException();
+                }
+
+                if (s_handlers.Count == 0 &&
+                    !Interop.Kernel32.SetConsoleCtrlHandler(s_handlerRoutine, addOrRemove: true))
+                {
+                    throw Win32Marshal.GetExceptionForLastWin32Error();
+                }
+
+                var token = new Token(signal, handler);
```

Comment:
Is the `Token` used to not root the `PosixSignalRegistration` from `s_handlers`?

Reply:
Right.